feat: honor `nativeTheme.themeSource = 'dark'` before creating BrowserWindow on Windows

ckerr opened the pull request:
#### Description of Change

Fixes #23479.

With this patch, setting `nativeTheme.themeSource = 'dark'` or `'light'` *before* creating the window is honored and the window will be rendered with the correct colors when shown.

As a side-effect of this change, the setting themeSource dynamically now works on Linux too.

There is still a wart: Changing themeSource dynamically when the window is already shown isn't perfect: the property is set, but the window still has to be poked before the change takes effect, e.g. changing focus or minimizing/restoring the window. The usual programmatic ways of forcing a window refresh don't seem to work here; perhaps this is my error but let's see about getting guidance for the right recipe to use.

This PR adds a lot of scaffolding to pull in the undocumented Win32 dark mode APIs. Much of this ground is covered by https://github.com/ysc3839/win32-darkmode, which is license-compatible (MIT).


TODO:
- [x] Can the titlebar issue described above be fixed? 
- [x] `win\dark_mode.cc` is really big. Audit it to see if any can be removed

CC @MarshallOfSound  @electron/wg-releases 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Improved dark mode support on Windows.

MarshallOfSound reviewed (changes requested):
This is a tricky one because it "fixes" a thing that I would argue isn't expected behavior.  On Windows there are maybe 3 apps that do this and all of them are windows built-in.  The private API this utilizes to do this is super hacky, apparently depends on exact versions / builds of Windows and in theory could break at any time.

If there isn't a public API for this maybe we should ask for one? It doesn't feel right adding this amount of private API complexity to do a thing that by most definitions isn't expected behavior (and speaking for one app in particular we'd want to opt out of)

ckerr commented:
@MarshallOfSound it looks like we'll be using undocumented API if we want this to work in the short term. I agree this is a lot of code for a small amount of functionality -- but at the same time, I think having this feature work is still worthwhile. What if I shave down all the code to only the minimum subset that we need, and then we see how large that patch is?

MarshallOfSound commented:
@ckerr I'm _really_ not a fan of this much private API usage.  If this were to land against my recommendations the minimum requirements would be:

* We've done our damndest to reduce the private API usage by as much as possible, the smaller the surface area the better
* The code should at least be vaguely readable, it's almost there atm but it's still a lot of `_` and magic enums
* This feature should 100% be disabled by default (opt-in) and no code path should call these APIs unless the individual app has opted in

samuelmaddock commented:
If this feature were to land, would it allow apps to forcefully apply dark mode? As opposed to applying it only when the user has chosen that OS theme.

Aside from the issue of private API usage, this would be a nice addition for apps which are primarily dark in aesthetics, but aren't using a frameless window.

ckerr commented:
> * We've done our damndest to reduce the private API usage by as much as possible, the smaller the surface area the better

I've removed everything not necessary to fix the titlebar issue reported in #23479. This revised patch only loads enough API to determine how to see if dark mode is supported, if it's allowed, and how to set it for newly-created windows. dark_mode.cc is now about 60% smaller despite having new comments to address the next point --

> * The code should at least be vaguely readable, it's almost there atm but it's still a lot of `_` and magic enums

I've tried to eliminate confusing code where possible, added explanatory comments, limited variables' scope, and moved variables to be declared just before use. Looking up function pointers from the dlls is still verbose -- you have to declare the function type, then declare a variable, and then do the lookup -- but I've added some template helper code to make it more readable.

> * This feature should 100% be disabled by default (opt-in) and no code path should call these APIs unless the individual app has opted in

I'm not  sure this is a good idea -- it doesn't seem productive to require users to opt-in to a bugfix? This fixes an existing code in our dark mode:

```js
nativeTheme.themeSource = 'dark';
new BrowserWindow(...)
// windows' titlebar will not be dark

ckerr commented:
> If this feature were to land, would it allow apps to forcefully apply dark mode? As opposed to applying it only when the user has chosen that OS theme.

Yes it would. The use case is literally setting `nativeTheme.themeSource = 'dark';` before creating a window. :)

MarshallOfSound commented on buildflags/buildflags.gni:
```diff
@@ -33,4 +33,7 @@ declare_args() {
   enable_builtin_spellchecker = true
+
+  # Undocumented Windows dark mode API
+  enable_win_dark_mode = true
```
Is this supposed to be enabled by default?

ckerr replied:
Disbled by default in 8508c62af5a9015dfebc9d38d68486a763b70aa6

MarshallOfSound commented on buildflags/BUILD.gn:
```diff
@@ -20,6 +20,7 @@ buildflag_header("buildflags") {
     "ENABLE_ELECTRON_EXTENSIONS=$enable_electron_extensions",
     "ENABLE_BUILTIN_SPELLCHECKER=$enable_builtin_spellchecker",
     "ENABLE_PICTURE_IN_PICTURE=$enable_picture_in_picture",
+    "ENABLE_WIN_DARK_MODE=$enable_win_dark_mode",
```
Can we call it something more specific to this feature.  i.e. `ENABLE_WIN_DARK_MODE_WINDOW_UI` or something

ckerr replied:
Renamed in 42214684722c2aad1548ee539aa2bf65046d009a

MarshallOfSound commented on typings/internal-ambient.d.ts:
```diff
@@ -24,6 +24,7 @@ declare namespace NodeJS {
     isPictureInPictureEnabled(): boolean;
     isExtensionsEnabled(): boolean;
     isComponentBuild(): boolean;
+    isWinDarkModeEnabled(): boolean;
```
```suggestion
    isWinDarkModeWindowUiEnabled(): boolean;
```

ckerr replied:
Fixed 2d54188d5c8e15a9d9474cff9d23cb10679c6fe1

MarshallOfSound reviewed (approved):
One missed replacement, other than that :shipit:

deepak1556 commented on shell/browser/win/dark_mode.cc:
```diff
@@ -0,0 +1,197 @@
+void InitDarkMode() {
+  HMODULE hDwmApi = LoadLibrary(L"DWMAPI");
+  if (hDwmApi != 0) {
+    Symbol(hDwmApi, &_DwmSetWindowAttribute, "DwmSetWindowAttribute");
```
Can you use `base::win::PEImage` instead of defining our own symbol import utility.

ckerr replied:
refactored to use `base/` in 46ec6e8a50207a8c62a8e329d4e3131168c2095d

ckerr commented on shell/browser/win/dark_mode.cc:
```diff
@@ -0,0 +1,197 @@
+
+#include "shell/browser/win/dark_mode.h"
+
+#pragma comment(lib, "Uxtheme.lib")
```
moved to deps in da3108c868907c7c3daff5431aba6c984b9cdcbe

deepak1556 commented on shell/browser/win/dark_mode.cc:
```diff
@@ -0,0 +1,195 @@
+  DWORD major = 0, minor = 0;
+  RtlGetNtVersionNumbers(&major, &minor, &g_buildNumber);
+  g_buildNumber &= ~0xF0000000;
+  if (major != 10 || minor != 0 || !CheckBuildNumber(g_buildNumber)) {
```
Can you use `base::win::GetVersion` instead, this seems like you will have to update `CheckBuildNumber` for every OS release.

ckerr replied:
Yes for certain we can use `base::win::OSInfo` instead, that's a great idea.

But... yes, absolutely there will be upkeep on this because there is still no documented API for this and it could all change in a future version of Windows. See for example how ordinal 135 in uxtheme changed from `AllowDarkModeForApp()` to `SetPreferredAppMode()` in version 1903. It's better for dark mode to just stop working in newer versions than it is to potentially call an unknown function signature.

Until there is supported stable API for these calls, I don't see a safe alternative.

ckerr replied:
Migrated build/version checks to `base::win::OSInfo` in b1d76af6d4f386c06e871bbe27e6af3c004b2f68

gpetrov commented:
Can we have this in production now without the flag?
